packages/react-native-web/src/exports/Image/index.js

```diff
@@ -353,6 +296,85 @@ ImageWithStatics.queryCache = function (uris) {
   return ImageLoader.queryCache(uris);
 };
 
+/**
+ * Image loading/state management hook
+ */
+const useSource = (callbacks, source: ?Source) => {
+  const [resolvedSource, setResolvedSource] = React.useState<ImageSource>(() =>
+    resolveSource(source)
+  );
+
+  const [status, setStatus] = React.useState(() =>
+    ImageLoader.has(resolveSource.uri) ? LOADED : IDLE
+  );
+
+  const [result, setResult] = React.useState<ImageSource>(resolvedSource);
+
+  // Trigger a resolved source change when necessary
+  React.useEffect(() => {
+    const nextSource = resolveSource(source);
+    setResolvedSource((prevSource) => {
+      // Prevent triggering a state change if the next is virtually the same as the last loaded source
+      if (JSON.stringify(nextSource) === JSON.stringify(prevSource)) {
+        return prevSource;
+      }
+
+      return nextSource;
+    });
+  }, [source]);
+
+  // Always use the latest value of any callback passed
+  // By keeping a ref, we avoid (re)triggering the load effect just because a callback changed
+  // (E.g. we don't want to trigger a new load because the `onLoad` prop changed)
+  const callbackRefs = React.useRef(callbacks);
+  callbackRefs.current = callbacks;
```

Comment:
Thanks for the explanation here. I guess we cannot do much about people passing anonymous functions to an `Image` component. Though it sounds like you're saying that would trigger a new "load" of the image?

Reply:
> I guess we cannot do much about people passing anonymous functions to an `Image` component. Though it sounds like you're saying that would trigger a new "load" of the image?

Yes if something causes a re-render an inline function declaration would be considered a new callback 
This maybe wasn't a big problem in the past, because RN web simply rendered an `<img src=` with the same `src` as the one being loaded (unless default source is involved)
Because of the rendered `<img src=` the browser initiated a request, and unless the `src` changes to `null` it wouldn't stop it (2 or more calls to the `ImageLoader.load` didn't matter much)

But for requests made with headers we know this works differently - it would abort the only `fetch` request downloading the image, so we don't want to be disrupted by an untimely or unintended `onLoad` prop change

(This is covered in both this PR and the new PR and even in 2 different ways 😂, albeit both aren't as straightforward as I want them to be)